### ISSUE TITLE
v5 No upstream DNS response acknowledged by pihole

### ISSUE THREAD

**Issue by Alphaprot** [pi-hole/pi-hole#3474]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`1`

---
**Expected behaviour:**

`pi-hole should be able to receive an answer from the upstream DNS servers (in my case 1.1.1.1 and 1.0.0.1 by cloudflare) and thus serve the client's request`

**Actual behaviour:**

`pi-hole seems to be able to send a request to an upstream DNS server but does not get a reply, which causes name resolution to fail after the time limit is exceeded. Firewall logs show allowed egress to the correct servers (e.g. 1.1.1.1:53), but nothing gets back.`

**My setup**

`Three major devices in the setup:

- ISP router is a FRITZ!Box 6490 cable modem. It connects via Ethernet to an OPNsense firewall. No port forwarding is currently enabled, because it shouldn't be necessary (the ISP router uses stateful packet inspection, thus everything outbound from behind the NAT should be allowed to pass in the opposite direction).

- OPNsense is configured to use pi-hole as the DNS server, furthermore it blocks every other DNS-requests trying to bypass pihole. Can confirm this setup working, at least it worked before. Pi-hole itself is located in my management VLAN, a firewall rule allows however access to pi-hole on port 53 from any other local VLANs.

- Pi-hole servers two purposes: being the traditional DNS forwarder to filter out ads and malicious content and local DNS resolving (which is configured via the pihole GUI). Pi-hole has DNSSEC enabled, however, by disabling it things won't change. `

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

`Because the pi appears to be unable to connect to the Internet at all (ping requests to ip-addresses result in 100% packet loss, too, because no ICMP  packets are coming back),

So, again time for pastebin: [Link](https://pastebin.com/44cX1P9W)`

**Troubleshooting undertaken, and/or other relevant information:**

`After monitoring my OPNsense logs, it appears to me that no responses arrive at the pihole. They won't be blocked; there is no indication of that in the firewall logs.
Which makes the whole thing even stranger to me - a dns-lookup from the firewall-host (OPNsense) itself works; but only using its own DNS after the pihole lookup times out.`

**Comment by PromoFaux** [comment]
If you put the Pi-hole device into a DMZ, or atleast beyond the firewall, does it work then?

My gut tells me that something on your firewall is set up too strictly (not an area I have any expertise in), as this is not normal behaviour for Pi-hole.

I guess networking must have been working at some point for Pi-hole to even be installed on the device :)

**Comment by Alphaprot** [comment]
Aaah, well - I am prone to thinking to complicated in problem solving. If bypassing the firewall completely (simply by connecting the pi-hole to the network of the ISP router), I am able to ping (name resolution on the pi also working again.).

Seems odd, since I haven't changed anything in the firewall ruleset itself, but it appears to me that something in the interconnection of the ISP router and the firewall's WAN interface might be misconfigured..

That's just a feeling though, I'll look into that. So, Adam - even though I probably disturbed you once again for complete unneccessary reasons - thanks for the hint in problem solving :)

**Comment by PromoFaux** [comment]
Everybody gets one... (or two 😉)!

Good luck...

**Comment by Alphaprot** [comment]
If you are like me, you'd be curious why something failed rather than "It failed. Period.":

So, yep, you were right - but the problem wasn't my ruleset (which is still perfectly fine) but rather outbound NAT (or the lack of it, to be more precise):
it appeared to me that the exposed-host setting wasn't working before, so my firewall still double NATted behind my ISP router. To counteract that, I turned of outbound NAT on the firewall.

I looked into that misconfig which prevented the exposed host feature (and thus bypassing of the ISP modems firewalling and NAT), found out, that there were to identical entries (and the wrong one was selected). After changing it, nothing worked.

Interestingly enough no blocking of inbound DNS replies was shown in the firewall view - but it was shown in the status table (remember, it's a stateful firewall, so it keeps track of everything that goes out to enable solicited replies to get back throug the firewall). Turned Outbound NAT on - and voila...

So - yep you were right, no failure of pihole.
